### DISCUSSION TITLE
Agent-based traffic simulation model (Mesa + UXsim)

### DISCUSSION THREAD

**Discussion by EwoutH** [toruseo/UXsim#101] [Show and tell]
I'm developing a hybrid Agent-based traffic simulation model, using [Mesa](https://github.com/projectmesa/mesa) en UXsim. It's based on a real-world road network extracted and processed with [OSMnx](https://github.com/gboeing/osmnx). It's in very early stages, but development can be followed here:

- https://github.com/EwoutH/urban-self-driving-effects/tree/main/model

The initialisation of the UXsim world is in [traffic.py](https://github.com/EwoutH/urban-self-driving-effects/blob/main/model/traffic.py), and adding demand, retrieving travel times and executing the simulation will happen in [model.py](https://github.com/EwoutH/urban-self-driving-effects/blob/main/model/model.py) (mainly in the `step` function).

First two TODOs:
- Add demand from agent mode choices (to regions)
- Get travel times back between areas to update expected car travel time

Once it's in further stages I will describe it more extensively, and probably open a PR to add it as an example. But for now a quick share for anyone interested to follow.

CC @toruseo 

![self-driving-research-flow-Conceptual-models-v1 drawio (1)](https://github.com/user-attachments/assets/82486773-0ac6-4e64-b78d-ac5a2209e551)
_Early conceptual model_

**Comment by toruseo** [comment]
Thanks for sharing, this is an ambitious project!

> Get travel times back between areas to update expected car travel time

If you need travel time only (not route), you may use `World.ROUTECHOICE.dist[i,j]` that store the current instantaneous travel time from node i to j.

This may be out of the scope, but it would be possible to formulate this as a day-to-day model, in which the "Travel time & price lookup" is came from previous day's experience so that it does not require significant external variables.

**Reply by EwoutH** [reply]
Thanks!

> If you need travel time only (not route), you may use World.ROUTECHOICE.dist[i,j] that store the current instantaneous travel time from node i to j.

This would be useful as a starting point! Maybe we can keep a second counter like `historical_dist` which keeps the latest (few) travel time(s) for each node combination of journeys actually completed.

> would be possible to formulate this as a day-to-day model, in which the "Travel time & price lookup" is came from previous day's experience so that it does not require significant external variables.

If both computational time and development time allow it, this is exactly what we plan on doing.

**Comment by EwoutH** [comment]
Quick update: Model is running and has reached minimal-viable product status, with a latest interaction bug between Mesa and UXsim resolved (correct bookkeeping of time in both is important): https://github.com/EwoutH/urban-self-driving-effects/commit/2a702c36c916b88658edae16a8997fb6408e8ef5

The modal is simulating traffic in a ~half a million city during a full day, with up to 50 thousand vehicles simultaneously in UXsim. Each 900 seconds the model recalculates all travel times between all nodes on which allows Agents to take new mode choice decisions.

First traffic patterns are following expectations. Lot's of "magic numbers" of course, tomorrow is data collection, aggregation, plotting and validation day. Goal is to determine where to refine more.

Then it's taking a step back to the original research goals to see which questions we can exactly answer, what we need to be able to do that, which are unfeasible.

**Comment by EwoutH** [comment]
Draft paper: https://github.com/EwoutH/urban-self-driving-effects/blob/main/paper/paper.md

@toruseo I hope that I explained the gist of UXsim well!

**Reply by toruseo** [reply]
Great work! And nice title: "Autonomous vehicles don't replace cars: they replace buses and bikes". I think AV fleet can be a good public transport if managed properly.

The description on UXsim is very good and fair. I didnt expect that UXsim is applicable to such a comprehensive transportation analysis. This is really encouraging work. Thanks!

**Comment by toruseo** [comment]
@EwoutH 
I happen to find that you are preparing to submit [your work to TR-C](https://github.com/EwoutH/urban-self-driving-effects/blob/main/paper/trc_paper.md). I am glad to hear that you have achieved such great results!

BTW, the [terms of use of UXsim](https://github.com/toruseo/UXsim#terms-of-use--license) state below
> When publishing works based on UXsim, please cite:
> 
> - Toru Seo. [Macroscopic Traffic Flow Simulation: Fundamental Mathematical Theory and Python Implementation](https://toruseo.github.io/misc/MacroTrafficSim_English_summary.pdf). Corona Publishing Co., Ltd., 2023.
> - Toru Seo. [UXsim: An open source macroscopic and mesoscopic traffic simulator in Python-a technical overview](http://dx.doi.org/10.48550/arXiv.2309.17114). arXiv preprint arXiv: 2309.17114, 2023

So could you please cite these ones?

**Reply by EwoutH** [reply]
Certainly! If the JOSS paper is accepted at the time, would you like me to cite that one?

**Reply by toruseo** [reply]
Thanks. That would be great!
